### PR TITLE
Edit profile

### PR DESCRIPTION
**Pull request by okg00dbye**
Changes
- Create EditProfile page
- Add RPC and backend methods for edit profile
- Add + link EditOwnProfile in frontend
- Style the profile pages (albeit incomplete D: )

TODO:
- style
- allow user to select avatar ahh couldn't get vant popup to work /flip table/
- can't get semicircle to be wHITE /flip chair/
- verify that edit profile works

Sorry this is not fully complete but maybe can merge first, I think imma switch over to work on gem logs RPCs
image
image

### CODE REVIEW

**Comment by sebastiantoh** [comment]
why `can't get semicircle to be wHITE /flip chair/`? Isn't it just setting the style

File: frontend/src/components/SemiCircleBg.vue [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<template>
+  <div class="stack-container">
+    <div class="semi-circle" v-bind:style="color" />
```
Comment (sebastiantoh) [inline frontend/src/components/SemiCircleBg.vue]: i think the RHS of `v-bind:style="color" ` must be some kind of pseudo css. e.g. `{ backgroundColor: color }`

File: backend/internal/adapters/primary/handlers/internal/userservice.go [hunk 2]
```diff
@@ -196,3 +196,29 @@ func (s *userService) DeclineFriendRequest(
...
+func (s *userService) EditOwnProfile(
+	c context.Context,
+	r *usersapi.EditProfileRequest,
+) (*emptypb.Empty, error) {
+	userID, ok := c.Value(contextkeys.UserID).(int64)
+	if !ok {
+		return nil, status.Error(codes.Internal, "unable to retrieve user ID from context")
+	}
+
+	if r.GetNewAvatar() != nil {
```
Comment (sebastiantoh) [inline backend/internal/adapters/primary/handlers/internal/userservice.go:209]: nit, which you can feel free to ignore: i don't think you should split the use cases here. It should be a single use case, then in the repositories, you would perform 2 updates (in a single transaction). That way you won't have one failing and the other succeeding.

File: backend/internal/adapters/secondary/repositories/user.go [hunk 3]
```diff
@@ -193,3 +193,29 @@ func (r *userRepository) DeclineFriendRequest(requesterID int64, receiverID int6
 
 	return nil
 }
+
+func (r *userRepository) UpdateAvatar(id int64, newAvatar []byte) error {
+	sql, _, _ := qb.Update("users").Where(goqu.C("id").Eq(id)).Set(goqu.Record{
```
Comment (sebastiantoh) [inline backend/internal/adapters/secondary/repositories/user.go]: i thnk goqu has issues with bytes. You may need to use prepared statements instead. Can take a look at the gem repository. I did something similar

File: frontend/src/pages/Profile/ProfileAvatar.vue [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  },
+  components: {
+    CircleAvatar,
+  },
+});
+</script>
+
+<style scoped lang="less">
+// TODO: abstract this out with friend-cell's gradient border maybe,
+// but not sure how to abstract CSS
```
Comment (sebastiantoh) [inline frontend/src/pages/Profile/ProfileAvatar.vue:21]: wdym by this? You can define a gradient varibale in `index.less` and reuse it here

Comment (okg00dbye) [reply frontend/src/pages/Profile/ProfileAvatar.vue:21]: can da gradient variable take in prop like radius

Comment (sebastiantoh) [reply frontend/src/pages/Profile/ProfileAvatar.vue:21]: Hmm, what's the radius for? You can define another radius variable, but it's not something the component can customise

Comment (okg00dbye) [reply frontend/src/pages/Profile/ProfileAvatar.vue:21]: the radius of the border is diff for FriendCell and ProfileAvatar

File: frontend/src/components/SemiCircleBg.vue (hunk 1 above)
Comment (okg00dbye) [reply frontend/src/components/SemiCircleBg.vue]: no wonder!

**Comment by okg00dbye** [comment]
> why `can't get semicircle to be wHITE /flip chair/`? Isn't it just setting the style

was da pseudo-css thingy, made it white and now it's invisible but at least it's white

File: backend/internal/adapters/secondary/repositories/user.go (hunk 3 above)
Comment (okg00dbye) [reply backend/internal/adapters/secondary/repositories/user.go]: updated, tq!

**Comment by okg00dbye** [comment]
TODO: save button shouldn't do anything if no changes have been made

File: frontend/src/pages/Profile/ProfileAvatar.vue (hunk 4 above)
Comment (sebastiantoh) [reply frontend/src/pages/Profile/ProfileAvatar.vue:21]: :o which line is that referring to? ps not super familiar with css haha

**Review by sebastiantoh — APPROVED** [review]
lgtm!

**Pull request merged**